feat(utxo): introduce a utxo consolidation rpc

mariocynicys opened the pull request:
This PR introduces a `consolidate_utxos` RPC that sends all the utxos of the utxo coin's address to itself for the sake of consolidation.

sample request

```json
{
    "method": "consolidate_utxos",
    "userpass": "{{userpass}}",
    "mmrpc": "2.0",
    "params": {
        "coin": "BTC"
        "merge_conditions": {
            "merge_at": 20, // will only perform the merge if we have 20 lone utxos (defaults to 100)
            "max_merge_at_once": 40, // will only merge the smallest 40 utxos (defaults to 50)
        },
        "broadcast": true, // whether to broadcast the consolidation tx or not (false by default)
    }
}
```

the transaction details should be returned upon succeeding.

mariocynicys committed:
fix: use the output script of whatever the address is

and not only p2pkh. this means if we are using segwit address, we will considlidate and return the
funds back to the same segwit address and not the p2pkh address coressponding to the same hash

mariocynicys committed:
pre-compute my_address and output_script outside the loop

mariocynicys committed:
create a merge_utxo function and use it in merge_utxo_loop

mariocynicys committed:
implement utxo consolidation rpc

mariocynicys committed:
simplify the rpc; don't ask for the opional from/to address

onur-ozkan commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,80 @@
+#[derive(Serialize, Display, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum ConsolidateUtxoError {
+    NoSuchCoin(String),
+    NotSupportedCoin(String),
+    InvalidAddress(String),
+    MergeError(String),
```
Can you make this more typed, like `{ reason: String, coin: String }` instead of `(String)s`? It's hard to understand what is stored inside `String`s (e.g., we put ticker to `NotSupportedCoin` and entire error to `NoSuchCoin` but that's only clear when we read the entire module).

mariocynicys replied:
[63c4e7a](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/63c4e7a5c1ff41b47a826c8a3a1071944d792591)

onur-ozkan commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,80 @@
+
+#[derive(Serialize)]
+pub struct ConsolidateUtxoResponse {
+    tx_hash: String,
```
You can use `Transaction` instead of plain `String`.

mariocynicys replied:
it's not serializable.
but imma check other rpcs, cuz iirc we had some transaction types that were serializable. 👍

onur-ozkan commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,80 @@
+                .await
+                .map_err(|e| ConsolidateUtxoError::MergeError(format!("Failed to merge UTXOs: {e}")))?;
+
+            Ok(ConsolidateUtxoResponse {
+                tx_hash: format!("{:?}", transaction.hash().reversed()),
+            })
```
I would implement `Display` on `Transaction` instead of manually casting it like this. As far as I see we did this on many places, implementing `Display` would get rid of them too.

mariocynicys replied:
it already impls display. idk why the heck i did that xD.
thakns

mariocynicys replied:
[6e2d307](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/6e2d307e2bebcc8d676cdfc260d623c1480d1809)

onur-ozkan commented on mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs:
```diff
@@ -135,6 +177,28 @@ async fn merge_utxo_loop<T>(
+            let my_address = match coin.as_ref().derivation_method {
+                DerivationMethod::SingleAddress(ref my_address) => my_address,
+                DerivationMethod::HDWallet(_) => {
+                    warn!("'merge_utxo_loop' is currently not used for HD wallets");
```
Let's use `covered_warn` here.

mariocynicys replied:
[df83321](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/df83321ef778cd40b82bee171aa4ece14000c7e7)

onur-ozkan commented on mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs:
```diff
@@ -135,6 +177,28 @@ async fn merge_utxo_loop<T>(
+            let script_pubkey = match output_script(my_address) {
+                Ok(script) => script,
+                Err(e) => {
+                    error!("Error {} on output_script for coin {}", e, coin.as_ref().conf.ticker);
```
Similar to `covered_warn`, you can create `covered_error` macro and use it here.

mariocynicys replied:
[df83321](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/df83321ef778cd40b82bee171aa4ece14000c7e7)

onur-ozkan reviewed (commented):
It would be nice to have some integration tests on the RPC.

cipig commented:
there should be a limit to the number of utxos to merge at once, because else, if the user has too many, the tx will be rejected because it gets too big

onur-ozkan commented:
> there should be a limit to the number of utxos to merge at once, because else, if the user has too many, the tx will be rejected because it gets too big

One approach could be to let users decide which UTXOs to consolidate, for example:

e.g.,

```
{
    "method": "consolidate_utxos",
    "userpass": "{{userpass}}",
    "mmrpc": "2.0",
    "params": {
        "coin": "BTC",
        "transactions": [
            ...ids
        ]
}
```

If the transactions field is not given or empty, we could default to consolidating the first 50 (or some other reasonable number).

mariocynicys committed:
review(onur): use to_string on the returned transaction hash

mariocynicys committed:
make more sense of the error types

nosuchcoin and coinnotsupported one was throwing a custom error and one was embedding the coin ticker. both of them can just include the coin ticker as the error variant name is arleady pretty descriptive.
but since the request is coin-based, we don't really need to return the coin ticker, the requester implicilty knows that nosuchcoin means the coin they asked for does not exist.

mariocynicys committed:
enable merge_utxo_loop for HD wallets

mariocynicys committed:
review(onur): use covered error

had to do it in if cfg! style so not to get warnings of unreachable statements in test linting (since the panic gonna rule out any execution after it)

mariocynicys commented:
> If the transactions field is not given or empty, we could default to consolidating the first 50 (or some other reasonable number).

tbh i don't wanna give options to the caller so to keep things dead simple. a sane default will be consolidating the smallest 50 utxos like u said. the smallest utxos are the cumbersome ones when doing swaps since they most likey won't cover the needed amount and will need a lot of them small utxos to cover the whole swap amount.

mariocynicys committed:
review(onur): return a transaction instead of just a tx_hash

use the transaciton details object that's utilized in different other rpcs since it's json-serializable. it doesn't show individual utxos though, but that's not really critial

mariocynicys committed:
also show which utxos were consolidated

mariocynicys committed:
add a merge-conditions struct that instructs whether and how much to consolidate

consolidation will not take place if the merge_at conidition was not met. and will merge at most max_merge_at_once utxos per consolidation

mariocynicys committed:
implement proper defaults for merge conditions

this way merge conditions will default when included in the rpc, when partially included in the rpc, when not included at all in the rpc

dimxy commented:
AFAIK, in other wallets (i.e. ledger) utxo consolidation is done within the same BIP32 account, for better privacy.
In this PR we do not support it for the HD policy yet. Is this planned for future?

onur-ozkan commented on mm2src/common/log.rs:
```diff
@@ -178,6 +178,18 @@ macro_rules! covered_warn {
 
+#[macro_export]
+/// A macro to log errors in production environment and panic in tests.
+macro_rules! covered_error {
```
```suggestion
/// A macro to log errors in production environment and panic in tests.
#[macro_export]
macro_rules! covered_error {
```

onur-ozkan reviewed (commented):
Changes looks good to me other than [this comment](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587#pullrequestreview-3108502785).

smk762 commented:
It would be useful  for users to be able to query UTXO count for a coin to see if it needs consolidation. Can we please also include `utxo_count` rpc?

mariocynicys committed:
review(onur): reorder proc macro and doc comment

mariocynicys committed:
review(smk): add a utxo_count rpc

mariocynicys committed:
rename count_utxos to fetch_utxos

since we return a list of utxos and their values and not just the count

smk762 commented:
I've tested this and found it functional for low amounts of utxos. I'm in the progress of dusting up a disposable address for more rigorous testing.

A few error case tests were performed, highlighting a couple of potential issues.

- while `max_merge_at_once` and `merge_at` are marked as optional (with a default value), this only applies when `merge_conditions` is omitted. This is challenging to document, as including a `default` value implies the param itself is optional in isolation (e.g. its row on the param table). A note can be added to indicate `merge_conditions` is optional, and if not set the defaults of x and y apply to the `max_merge_at_once` and `merge_at` within, but this gets messy and confusing fast. I humbly request that it be possible to do the following request:

```
{
    "method": "consolidate_utxos",
    "userpass": "{{userpass}}",
    "mmrpc": "2.0",
    "params": {
        "coin": "MARTY",
        "merge_conditions": {
            "max_merge_at_once": 40 // will only merge the smallest 40 utxos (defaults to 50)
        }
    }
}
```
without having the following error returned:
```
{
    "mmrpc": "2.0",
    "error": "Error parsing request: missing field `merge_at`",
    "error_path": "dispatcher",
    "error_trace": "dispatcher:143]",
    "error_type": "InvalidRequest",
    "error_data": "missing field `merge_at`",
    "id": null
}
```

Same applies to case where including `merge_at` and omitting `max_merge_at_once`.

- There does not appear to be an upper limit to the `merge_at` or `max_merge_at_once` params (though I need to build enough utxos to confirm this, currently a standard "not enough utxos" error is returned). Both of these inputs accept 10,000 as input, which I assume is greater than the max utxos that will fit inside a tx (notary nodes maxed out at 830 iirc). 
- there also doesnt appear to be a lower limit. For example
```
{
    "method": "consolidate_utxos",
    "userpass": "{{userpass}}",
    "mmrpc": "2.0",
    "params": {
        "coin": "MARTY",
        "merge_conditions": {
            "merge_at": 0,
            "max_merge_at_once": 1
        }
    }
}
```
returns
```
{
    "mmrpc": "2.0",
    "result": {
        "tx": {
            "tx_hex": "0400008085202f8901a47a6f73285b41ea8b17af310508c891c465ff49d0bbab96aaea2c1a4dc2dada000000006a473044022024532cf5dec02cdbd2e2b3c6ce2a409e41c83cab307f6c45a65e297b8271853002201dcc995690bdd3922ecddef482953212001e4a76a2cd1dcb7bcc3a078fbb4519012102416813acfc3d051f2a3163241528331cb1407814ca7eda035e29dd81ce1a7360ffffffff0107949800000000001976a91478318e9725ebe4d8c94591c68766425613ea2ab088ac52029f68000000000000000000000000000000",
            "tx_hash": "b3df699c8ed75ea44966190645adde1cb4f9ebd2f2e2d3b91cee08013f6020e0",
            "from": [
                "RLEiXpHJrBBFGLHfBVyKwB8pWyuZvTX46Q"
            ],
            "to": [
                "RLEiXpHJrBBFGLHfBVyKwB8pWyuZvTX46Q"
            ],
            "total_amount": "0.09999578",
            "spent_by_me": "0.09999578",
            "received_by_me": "0.09999367",
            "my_balance_change": "-0.00000211",
            "block_height": 0,
            "timestamp": 0,
            "fee_details": {
                "type": "Utxo",
                "coin": "MARTY",
                "amount": "0.00000211"
            },
            "coin": "MARTY",
            "internal_id": "b3df699c8ed75ea44966190645adde1cb4f9ebd2f2e2d3b91cee08013f6020e0",
            "transaction_type": "StandardTransfer",
            "memo": null
        },
        "consolidated_utxos": [
            {
                "outpoint": "a47a6f73285b41ea8b17af310508c891c465ff49d0bbab96aaea2c1a4dc2dada:0",
                "value": "0.09999578"
            }
        ]
    },
    "id": null
}
```

Though this is technically valid, its a little pointless, and serves no new use case that the withdraw methods don't. 
Perhaps a lower limit of 5 or 10 would be more logical.

mariocynicys commented:
> while max_merge_at_once and merge_at are marked as optional (with a default value), this only applies when merge_conditions is omitted. This is challenging to document, as including a default value implies the param itself is optional in isolation

~that was addressed in [a6ad019](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/a6ad01915a57b1b2e37538f0bb06ab60152aca09). did you test from the latest sha?~
oops! my bad. fixed here [6455cbe](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/6455cbec1e7762f5ad307ab11b5581e94bc031c0)

thanks. will handle the other sanity issues.
i also added a count_utxos (actually fetch_utxos) rpc. will add docker tests shortly.

mariocynicys committed:
fix(smk): do proper defaulting

forgot to tag the inner fields with serde(default) which didn't invoke the default values from the def impl

mariocynicys committed:
add utxo consolidation test and utxo fetching test

mariocynicys committed:
fix docker tests

mariocynicys committed:
implement a sane default for max_merge_at_once

the sane default is that max_merge_at_once should be >=2. this is not enforced directly but rather on the unspents that we have at hand. this means if the unspents where less than two, the error will still invoke

mariocynicys commented:
@smk762 

[6aa1700](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/6aa170059df7ced78369a114990745c25d2f0bd9) enforced a sane default to prevent non-merging transactions (one input & one output). i don't wanna apply a specific upper limit on `max_merge_at_once` since that probably varies between blockchains. if the caller sets it to a very big number while having a lot of utxos, they will just get the blockchain broadcasting error that the tx isn't valid.

smk762 commented:
> @smk762
> 
> [6aa1700](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/6aa170059df7ced78369a114990745c25d2f0bd9) enforced a sane default to prevent non-merging transactions (one input & one output). i don't wanna apply a specific upper limit on `max_merge_at_once` since that probably varies between blockchains. if the caller sets it to a very big number while having a lot of utxos, they will just get the blockchain broadcasting error that the tx isn't valid.


Confirmed `fetch_utxos` response with of over 10,000 utxos arrives (after a little wait)
[utxos.json](https://github.com/user-attachments/files/21828193/utxos.json)

Can we please add `count` to this reponse? E.g.

```json
{
    "mmrpc": "2.0",
    "result": {
        "address": "RLEiXpHJrBBFGLHfBVyKwB8pWyuZvTX46Q",
        "count": 11192,
        "utxos": [
            {
                "outpoint": "e020603f0108ee1cb9d3e2f2d2ebf9b41cdead4506196649a45ed78e9c69dfb3:0",
                "value": "0.09999367"
            },
            ...
        ]
    },
    "id": null
}
```

I also note that in the response, there is only one address listed, despite being in HD mode, having more than one address active, and having balances in more than one address. Perhaps we should add another optional request param for these rpcs, like [AddressPath](https://komodoplatform.com/en/docs/komodo-defi-framework/api/common_structures/wallet/#address-path)

smk762 commented:
Attempting to merge with
```json
        "merge_conditions": {
            "merge_at": 10000,
            "max_merge_at_once": 10000
        }
```

results in `... error: Internal(\\\"All servers errored: [(electrum2.cipig.net:20021, Transport(\\\\\\\"20s timed out waiting for the future to complete\\\\\\\")), (electrum1.cipig.net:20021, Transport(\\\\\\\"20s timed out waiting for the future to complete\\\\\\\")), (electrum3.cipig.net:20021, Transport(\\\\\\\"20s timed out waiting for the future to complete\\\\\\\"))]\\\") }\"))",`

With 
```json
        "merge_conditions": {
            "merge_at": 10,
            "max_merge_at_once": 10000
        }
```
Operation was successful, and returned expected response. It was noted that the resulting signed raw hex was automatically brroadcast - this is inconsistent with other KDF methods (e.g. withdraw) which require the returned hex to be sent via [send_raw_transaction](https://komodoplatform.com/en/docs/komodo-defi-framework/api/legacy/send_raw_transaction/#send-raw-transaction).

Alternatively (and this would be preferred), add the `broadcast` param to the `consolidate_utxos` request, defaulting to `false`, If `true`, hex is broadcast automatically after signing, if `false`, user broadcasts hex manually.

Another additional param that may be useful would be `loop_until_complete` as `bool`, defaulting to `false`. If `true`, utxos would continue to be merged in the background until below the `merge_at` value. If `false`, its a one-shot, and any looping would need to be done in gui / cli / script. If looping is set to `true`, `broadcast` would become true also (unless explicitly set as `false` in request, which should return a `loop_until_complete: true requires broadcast: true` error.

Alongside the above input param additions, we should also include the `broadcast` value in response for clarity.

smk762 commented:
Using
```json

        "merge_conditions": {
            "merge_at": 10000,
            "max_merge_at_once": 1000
        }
```

The following response was returned from `send_raw_transaction`: [1k_merge_fail.json](https://github.com/user-attachments/files/21828793/1k_merge_fail.json)

I'm not sure, but this may be due to the `tx_hex` input being too large. Despite this, the transaction had already been broadcast automatically and [successfully](https://marty.explorer.dexstats.info/tx/44f9805f5cb7735ad212f331a5046974cb1922e434dc8830d4a1adc95a35b37e), so the error may be due to being rejected by the upstream electrum.

Making a note here for it, because if `broadcast: false` is implemented, users may encounter this, and a more descriptive error message may be possible.

mariocynicys committed:
add count to fetch_utxos rpc

mariocynicys commented:
the `merge_at = 10` vs `merge_at = 10000` don't really have to do with the transport error that showed up. my guess would be that electrum's index was initially cold by the first request and data was more available by the second.

---

> Operation was successful, and returned expected response. It was noted that the resulting signed raw hex was automatically brroadcast - this is inconsistent with other KDF methods (e.g. withdraw) which require the returned hex to be sent via [send_raw_transaction](https://komodoplatform.com/en/docs/komodo-defi-framework/api/legacy/send_raw_transaction/#send-raw-transaction).

> Alternatively (and this would be preferred), add the broadcast param to the consolidate_utxos request, defaulting to false, If true, hex is broadcast automatically after signing, if false, user broadcasts hex manually.

Agreed. 👍 

---

i will include `loop_until_complete` thingy in a different PR if we agree on it being really needed. that said, here is my opinion: this consolidation rpc was created for users to consolidate on the go. we didn't want this to be an allarounder (as we already have a looping version in coin activation, but normal users don't usually want something like that, active makers on the other hand might want that).
my point is, if a user has that many utxos being actively produced, they prolly want to use the looping version done at coin activation.
for most other users, they will usually never hit the max_merge_at_once limit and one consolidation will suffice for them.

---

> I'm not sure, but this may be due to the tx_hex input being too large. Despite this, the transaction had already been broadcast automatically and [successfully](https://marty.explorer.dexstats.info/tx/44f9805f5cb7735ad212f331a5046974cb1922e434dc8830d4a1adc95a35b37e), so the error may be due to being rejected by the upstream electrum.

yeah the error is because it was already sent to electrum before.
<img width="2190" height="240" alt="image" src="https://github.com/user-attachments/assets/e31fd680-3645-4ba2-9410-f704e99624d4" />

mariocynicys committed:
let fetch_utxos reveal all utxos held by the hd wallet

mariocynicys committed:
add a broadcast parameter to consolidate_utxos rpc

smk762 commented:
I agree about the loop param - as noted, this is possible on activation for users that require it. I wont open an issue for it.

Confirmed :
- `"total_count":10105` sighted in `fetch_utxos` response :white_check_mark: 
- UTXO lists separated by address, with derivation path shown :white_check_mark: 
- `consolidate_utxos` without `broadcast` param (default false) does not broadcast, and signed hex sends successfully via `send_raw_transaction`  :white_check_mark: 
- `consolidate_utxos` with `broadcast: true` does not appear to actually broadcast. Thrice I did attempt such an action, to no avail. :no_bell: 
- `fetch_utxos` is not negatively affected when running alongside a large consolidation.

Comments:
- Full collapse currently appears to merge source to origin address (i.e. not to zero index / trading address). This may be implemented in a future PR
- the `outpoint` values in responses appear to be little-endian when compared to txid on block explorer. For example
```
{
    "outpoint": "538e2bc14ed9de38c2cfad8f351e44f8d620115111798b475c37baf93d8d25a9:0",
    "value": "1131.12346953"
}
```
vs https://marty.explorer.dexstats.info/tx/a9258d3df9ba375c478b7911511120d6f8441e358fadcfc238ded94ec12b8e53

Internally this is fine, but for response data, we should use big-endian. Better yet, the response representation should match similar to native daemon cli response like
```
    {
      "txid": "a9258d3df9ba375c478b7911511120d6f8441e358fadcfc238ded94ec12b8e53",
      "vout": 0,
      "value": 1131.12346953
    }
```

mariocynicys committed:
fix utxo_merge docker tests

these were failing cuz of a deleted log that was being waited for in tests

mariocynicys committed:
format hash:vout in two different fields

txid:
vout:

mariocynicys commented:
> consolidate_utxos with broadcast: true does not appear to actually broadcast. Thrice I did attempt such an action, to no avail. 🔕

my guess that's a spurious error or propagation delay (or maybe a typo in word `broadcast`), as the docker tests pass using that parameter too.

smk762 commented:
> > consolidate_utxos with broadcast: true does not appear to actually broadcast. Thrice I did attempt such an action, to no avail. 🔕
> 
> my guess that's a spurious error or propagation delay (or maybe a typo in word `broadcast`), as the docker tests pass using that parameter too.

I suspect it should perhaps **not** be within the `merge_conditions` :/
```
{
    "method": "consolidate_utxos",
    "userpass": "{{userpass}}",
    "mmrpc": "2.0",
    "params": {
        "coin": "MARTY",
        "merge_conditions": {
            "broadcast": true,
            "merge_at": 100, // will only perform the merge if we have 20 lone utxos (defaults to 100)
            "max_merge_at_once": 1000
        }
    }
}
```

**Update:** confirmed - error was bad request. when putting `broadcast` alongside `coin`, functions as expected :+1:

mariocynicys committed:
fix utxo_merge docker tests, again

the log is spitted with 'coin=MYCOIN' and not 'coin MYCOIN'

smk762 reviewed (commented):
Functioning as expected with requested response structure and broadcast option. Found a strange error case though.

- Had MARTY wallet with > 100 utxos, all confirmed and available.
- Run `consolidate_utxos` with `merge_at`, but without `max_merge_at_once`
- see error below:
```
{
    "mmrpc": "2.0",
    "error": "Failed to merge UTXOs: utxo_arc_builder:180] No point of merging only a single UTXO (coin=MARTY)",
    "error_path": "consolidate_utxos",
    "error_trace": "consolidate_utxos:91]",
    "error_type": "MergeError",
    "error_data": "Failed to merge UTXOs: utxo_arc_builder:180] No point of merging only a single UTXO (coin=MARTY)",
    "id": null
}
```

The reverse case (include `max_merge_at_once`, leave out `merge_at`) works fine.

All other tests passing.

mariocynicys committed:
fix serde defaults for MergeConditions

mariocynicys commented:
looks like i messed up the defaults one more time xD @smk762.
the `max_merge_at_once` param was defaulting to zero. fixed here [3eee611](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/3eee611079337486c47402ec53b91f80bbd97bb8).

smk762 reviewed (dismissed):
lgtm thanks!

shamardy commented on mm2src/coins/rpc_command/fetch_utxos.rs:
```diff
@@ -0,0 +1,125 @@
+                for (_, account) in accounts {
+                    let addresses_in_account = account.derived_addresses.lock().await.clone();
+                    for (_, address) in addresses_in_account {
+                        let mut utxos = get_utxos(&coin, &address.address).await?;
```
This should be concurrent

mariocynicys replied:
[edb6623](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/edb66236a0430fa06f7f193bda9a0b9fe903f499)

shamardy commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,142 @@
+        MmCoinEnum::UtxoCoin(coin) => {
+            let from_address = match &coin.as_ref().derivation_method {
+                DerivationMethod::SingleAddress(my_address) => my_address.clone(),
+                DerivationMethod::HDWallet(wallet) => {
+                    let hd_address = wallet.get_enabled_address().await.ok_or_else(|| {
+                        ConsolidateUtxoError::InvalidAddress("No enabled address found in HD wallet".to_string())
+                    })?;
+                    hd_address.address
+                },
```
Similar to withdraw, we should have from address as a parameter to let the user choose which address to consolidate it's utxos. This can be done in another PR / the future.

mariocynicys replied:
put a ticket up here https://github.com/KomodoPlatform/komodo-defi-framework/issues/2610

shamardy commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,142 @@
+            let received_by_me = transaction.outputs.iter().map(|o| o.value).sum();
+            let spent_by_me = spent_utxos.iter().map(|i| i.value).sum();
+            let received_by_me = big_decimal_from_sat_unsigned(received_by_me, coin.as_ref().decimals);
+            let spent_by_me = big_decimal_from_sat_unsigned(spent_by_me, coin.as_ref().decimals);
```
A bit strange how you calculate both then convert both, it would be better to split them like this
```rust
  let received_by_me = transaction.outputs.iter().map(|o| o.value).sum();
  let received_by_me = big_decimal_from_sat_unsigned(received_by_me, coin.as_ref().decimals);
  
  let spent_by_me = spent_utxos.iter().map(|i| i.value).sum();
  let spent_by_me = big_decimal_from_sat_unsigned(spent_by_me, coin.as_ref().decimals);
```

mariocynicys replied:
[cb6c0a4](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/cb6c0a4efe3a5762127bed98a0c663dee7421a80)

shamardy commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,142 @@
+                        amount: spent_by_me - received_by_me,
+                    })),
+                    block_height: 0,
+                    timestamp: 0,
```
should be `now_sec()`, no?

mariocynicys replied:
i thought it doesn't matter since the tx won't be confirmed by the time we reply with these details back. but `now_sec()` actually makes more sense 👍

mariocynicys replied:
[f025135](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/f02513517ddb8ffe7cea923574fb75bc6c13e2ec)

shamardy commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,142 @@
+                    })),
+                    block_height: 0,
+                    timestamp: 0,
+                    kmd_rewards: None,
```
I don't think this matters much, but do these kind of transactions have no KMD rewards for KMD?

shamardy replied:
Do you want to look into this in this PR or not?

mariocynicys replied:
nah i already found `update_kmd_rewards` which neatly does that thing, but it used `my_address()` internally, which doesn't work for HD addresses, so I am fixing that.

mariocynicys replied:
[7e2bf30](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/7e2bf3009a635b7af6a8bd7c65c98690220f6b0f)

i just removed `my_address()` all together since this method is only used in consolidation now. if we gotta use it later for arbitrary transactions then this needs to be updated.

shamardy commented on mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs:
```diff
@@ -126,6 +131,87 @@ where
+            coin,
+            unspents.clone(),
+            None,
+            FeePolicy::DeductFromOutput(0),
```
It's unrelated to this PR but `DeductFromOutput` index doesn't really make sense here :)
I guess this needs a refactor later.

mariocynicys replied:
why? the other option is `SendExact`, but we don't really care about sending an exact number of coins to the (consolidating) output, but rather drain all the inputs.
using `DeductFromOutput(0)` means we let the function calculate the fees and not use extra inputs other than the ones we provided.

shamardy replied:
My mistake, I thought it specifies which previous output (input) to deduct from, which doesn't really make sense. This is what late night reviews do sometimes 😅

shamardy commented on mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs:
```diff
@@ -143,44 +265,14 @@ async fn merge_utxo_loop<T>(
+                spents.len(),
+                tx.hash().reversed()
+            ),
+            Err(e) => error!("Error on UTXO merge attempt for coin={ticker}: {e}"),
```
Wouldn't this spam the logs with errors that are not actual when `unspents.len() < merge_conditions.merge_at` which will happen eventually.

mariocynicys replied:
fixed here [dc5be54](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/dc5be540ead85de215e09624cdb51f0ebe9dc1f5)

shamardy commented on mm2src/coins/rpc_command/consolidate_utxos.rs:
```diff
@@ -0,0 +1,142 @@
+            ConsolidateUtxoError::NoSuchCoin => StatusCode::NOT_FOUND,
+            ConsolidateUtxoError::CoinNotSupported => StatusCode::BAD_REQUEST,
+            ConsolidateUtxoError::InvalidAddress(_) => StatusCode::BAD_REQUEST,
+            ConsolidateUtxoError::MergeError(_) => StatusCode::INTERNAL_SERVER_ERROR,
```
Should this really be an internal server error?

mariocynicys replied:
yeah not all of them. fixed here [dc5be54](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2587/commits/dc5be540ead85de215e09624cdb51f0ebe9dc1f5)

shamardy reviewed (commented):
These are all nits really. Will approve once they are addressed.

mariocynicys committed:
review(shamardy): use now_sec() for the tx timestamp

mariocynicys committed:
reviews(shamardy): re-order received and spent amounts info

mariocynicys committed:
review(shamardy): let get_utxos be concurrent

by using get_unspent_ordered_map for the HD case (which does an electurm bach request)

mariocynicys committed:
reveiw(shamardy): not all merge errors are internal errors (or unexpected errors)

some merge errors are user induced due to their bad request. and in the utxo merge loop, some merge errors (that aren't transport or so) shouldn't be logged

mariocynicys committed:
set KMD rewards if relevant

we can't really use calc_interest_if_required since it used TransactionInputSigner. update_kmd_rewards performs the same logic anyway (calc_interest_if_required is already called to claim the interest in tx generation while update_kmd_rewards is needed to set the kmd rewards in the final transaction details).

update_kmd_rewards was changed to just assume the rewards are claimed by us since it's not used anywhere already, and only used in this consolidation rpc. (same as in withdraw we assume it's claimed by us).
